### DISCUSSION TITLE
Adding new interfaces to support hierarchical deployments

### DISCUSSION THREAD

**Discussion by drholmie** [testground/testground#1168] [General] [answered]
## Description

Adding the ability to use more interfaces for data network and create hierarchical deployments.
Currently two interfaces exists in the k8s runner. One for the control and the other for data. With this feature a user can add more interfaces to help in creating private and public networks in the same deployment. 

This feature would overall allow for flexible network architectures, and increased bench-marking capabilities as this can emulate more types of networks. 

## What defines this endeavour to be complete?
Being able to create and add docker containers to public and private networks. 

<!--Create a check list of what you expect to see as part of delivering on the ask--

- [ ] Item 1
- [ ] Item 2
- [ ] ...

**Comment by nonsense** [comment, marked as answer]
@drholmie could you elaborate a bit further on why you need multiple interfaces for the data network? What is it that you are trying to test, that is not possible by assigning a different subnet for various groups within the test case? If we have more information about your use-case maybe we can think of alternative solutions that we support today.

Theoretically it should be possible to add multiple interfaces for the data network with Weave (the networking solution we use for the data network within Kubernetes), but I remember hitting some roadblocks when trying to do that - I don't exactly remember if the issues were on a CNI level or with Weave or with the Weave CNI plugin... will look it up when I get a chance.

I think with the current CNI plugin that we use to attach multiple interfaces to pods in Kubernetes - https://github.com/cni-genie/CNI-Genie - it is only possible to attach 1 interface per real CNI plugin (i.e. flannel, or weave, or some other real, and not meta, CNI plugin). This might have changed over the last year, or there might be alternatives to CNI-Genie that we could use (such as https://github.com/intel/multus-cni), but I am fairly sure this was a limitation at the time when we implemented the control and data network.

**Comment by drholmie** [comment]
Thank you for the informative response! I'm working on a project at the moment regarding simulating different network architectures and wanted to know if using testground (or modifying it in some way) would suit our project. We'd like to have the capability of creating public and private networks by adding different interfaces. To provide an example, take two master nodes with two worker nodes each. Master nodes can ping each other. Workers can ping each other when under the same master node (including pinging their master node), but can't ping the workers under a different master node.   

Could you explain more on creating subnets within a test case? I don't have much experience with Go and did not know this was possible to add in a testcase :sweat_smile: 

From my (limited) experience I do think it is a CNI issue for weave. We were also looking into multus as well. What changes would be required to get that running with testground? We would be interested in helping out with development if the modification fits our use case better.

**Comment by nonsense** [comment]
> We'd like to have the capability of creating public and private networks by adding different interfaces. To provide an example, take two master nodes with two worker nodes each. Master nodes can ping each other. Workers can ping each other when under the same master node (including pinging their master node), but can't ping the workers under a different master node.

You might be able to do that if you split the nodes in different regions. See the https://github.com/testground/testground/tree/master/plans/splitbrain testplan as an example.

> Could you explain more on creating subnets within a test case?

The `splitbrain` and the [network/ping-pong](https://github.com/testground/testground/tree/master/plans/network) tests exercise that functionality from Testground.

---

This is documented at https://docs.testground.ai/traffic-shaping , but depends heavily on which runner you want to use: `local:docker` or `cluster:k8s` - `local:docker` is using docker networks (https://docs.docker.com/network/), whereas Kubernetes is setup to use Weave for the data network (16.0.0.0/4 CIDR). Documentation for `cluster:k8s` setup is at https://docs.testground.ai/runner-library/cluster-k8s/how-to-create-a-kubernetes-cluster-for-testground, and you can check the exact playbooks at https://github.com/testground/infra

> What changes would be required to get that running with testground?

Weave and CNI-Genie are setup while installing the cluster - https://github.com/testground/infra/blob/master/k8s/01_install_k8s.sh#L72-L75

If for example Multus (or another meta-CNI plugin) provides functionality that enables you to attach multiple inferfaces to pods, then it needs to be installed and configured on the cluster, and then the `sidecar` should also be updated - https://github.com/testground/testground/tree/master/pkg/sidecar . The `sidecar` is responsible for handling network configuration for testplan instances in Testground, and is coupled with the given runner (for example https://github.com/testground/testground/blob/master/pkg/sidecar/k8s_reactor.go is coupled with the exact Kubernetes setup using Weave and Flannel that we have in the `testground/infra` repo).

**Comment by drholmie** [comment]
Ah nice, thanks so much for the information and links! A few more questions. 
1. Are there any helper scripts to deploy this on non AWS infrastructure? 
2. With respect to the kubernetes cluster itself, is it possible to further extend or add plugins (CNI or otherwise), say for example running Jaeger or Linkerd, to collect more statistics, or adding OVS CNI plugins. Would testplans still function in this cluster? 
3. Also a little confused with how test plans are deployed using the kubernetes runner. (Please feel free to correct my misunderstandings 😅 ). Is the cluster something that we have to make on our own first (in a non AWS deployment), deploy it and then apply the testground daemons to it, where test plans would only take up as many resources as it should, for example using the ping pong test needs only two containers, so using it on the kubernetes runner only uses two pods (out of n pods deployed in the cluster) or does this deploy a kubernetes cluster with just two pods?

**Comment by nonsense** [comment]
1. Not at the moment, and we don't have this on our roadmap. However the `cluster:k8s` runner is based on Kubernetes, and we are using `kops` for our Kubernetes deployments, so it shouldn't be hard to use other than AWS infrastructure. We do have a bit of coupling to AWS, such as our use of EFS and EBS volumes, but I am pretty sure you can substitute this. (or just not use it).

2. Not sure what you mean with that. AFAIK Jaeger (https://www.jaegertracing.io/) is a tracing tool, not sure how it fits in the CNI realm. You can totally deploy it as a pod on you cluster though. You could use a different CNI such as Linkerd, but you might need to update the `sidecar` and the playbooks linked above.

3. Test plans are sent to the `cluster:k8s` runner within the Testground daemon, they are somewhat ortogonal to the Kubernetes deployment. When you run a Testground command, such a `testground run composition -f mycomposition.toml`, the test plan references from `mycomposition.toml` must be imported locally on the caller side, and the source code will be shipped to the Testground daemon.

> Is the cluster something that we have to make on our own first (in a non AWS deployment), deploy it and then apply the testground daemons to it

Yes, this is what the playbooks at https://github.com/testground/infra are for, but they are aimed only at AWS.

> for example using the ping pong test needs only two containers, so using it on the kubernetes runner only uses two pods (out of n pods deployed in the cluster)

The ping pong testplan would create 2 pods on Kubernetes, yeah. `(out of n pods deployed in the cluster)` - I am not sure what you mean with this. Kubernetes cluster has a limit of number of pods, and a limit of resources.

**Comment by drholmie** [comment]
>    However the `cluster:k8s` runner is based on Kubernetes, and we are using `kops` for our Kubernetes deployments, so it shouldn't be hard to use other than AWS infrastructure. 

Regarding non AWS I meant a personal kubernetes cluster (Not something hosted on the cloud), my bad should've clarified that :sweat_smile:  Let's say I already have a personal kubernetes cluster setup, would testground be able to run its deployments there?

>     2. Not sure what you mean with that. AFAIK Jaeger (https://www.jaegertracing.io/) is a tracing tool

Ah so here I meant is it possible to use other plugins to collect more stats too from the cluster if required.


> The ping pong testplan would create 2 pods on Kubernetes, yeah. `(out of n pods deployed in the cluster)` - I am not sure what you mean with this. Kubernetes cluster has a limit of number of pods, and a limit of resources.

Ah my bad, didn't explain this properly either :sweat_smile:, meant to ask, if a kubernetes cluster existed how would testground deploy testplans on that. 

Thanks again so much for all the help and clarifications : )

**Comment by nonsense** [comment]
@drholmie when running Kubernetes, I think the distinction is generally between `managed` and `self-managed` deployments. Testground is using `kops`, so we build a `self-managed` cluster. You can build the same `self-managed` cluster on variety of hardware - check kops for details - https://github.com/kubernetes/kops

Still as explained Testground is coupled with AWS at the moment, as we use features from EBS and EFS - you'd have to extract these and re-implement them for your purposes.

> Ah so here I meant is it possible to use other plugins to collect more stats too from the cluster if required.

Yes, you can do that. On the AMIs that we use (and those that are part of the infra playbooks), we install Netdata for example for monitoring. Nothing stops you from modifying the environment on Kubernetes as you see fit - you are free to install Jaeger for example :)

**Comment by nonsense** [comment]
@drholmie I converted your issue to a discussion, because it seems like you need a fairly unique setup, that might involve heavy refactor and updates on Testground side, that we don't currently have on the roadmap - specifically the support for multiple interfaces, that is limited with the current CNIs we use.

I think the Discussions section fits better this conversation - we can continue the conversation here, and I can provide more input if this is useful to you.

**Comment by drholmie** [comment]
Hi there @nonsense ! Thank you for converting this to a discussion. Regarding follow ups to this, my team has decided to try running Testground in a simpler p2p environment first. With this in mind we're currently aiming to run Testground on any kubernetes cluster (no dependency or support on AWS) with the prerequisite being the cluster must have kubernetes already set up.  
If this requires changes to Testground, we'd love to submit a PR on it : )

Thank you again so much for all the help! Will keep in touch regarding adding CNIs to Testground as well :)

**Comment by drholmie** [comment]
Hi there @nonsense! Happy new year! Not too sure if this is the right place to ask. In terms of a status update, got the [infra repo](https://github.com/testground/infra) scripts running on generic clusters (needs a bit of polish but it works :sweat_smile: ). However wanted to know if its possible to use different languages too, currently noticed that there's an sdk for js, is there any in the works for python? Is there also a way to call other scripts within a go testplan and take advantage of the testground apis without building a language specific sdk?
